pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -218,15 +256,73 @@ pub mod pallet {
         UnexpectedXCMResponse,
         /// Block production is pending for para id with successfully placed order
         BlockProductionPending,
+        /// Block production is not allowed for current slot
+        NotAllowedToProduceBlockRightNow,
+        /// Collator signature nonce is incorrect
+        IncorrectCollatorSignatureNonce,
+        /// Collator signature is invalid
+        InvalidCollatorSignature,
+    }
+
+    impl<T: Config> From<BuyingError<BlockNumberFor<T>>> for Error<T> {
+        fn from(value: BuyingError<BlockNumberFor<T>>) -> Self {
+            match value {
+                BuyingError::OrderAlreadyExists { .. } => Error::<T>::OrderAlreadyExists,
+                BuyingError::BlockProductionPending { .. } => Error::<T>::BlockProductionPending,
+                BuyingError::NotAParathread => Error::<T>::NotAParathread,
+                BuyingError::NotAllowedToProduceBlockRightNow { .. } => {
+                    Error::<T>::NotAllowedToProduceBlockRightNow
+                }
+            }
+        }
     }
 
     /// Proof that I am a collator, assigned to a para_id, and I can buy a core for that para_id
     #[derive(Encode, Decode, CloneNoBound, PartialEq, Eq, DebugNoBound, TypeInfo)]
-    #[scale_info(skip_type_params(T))]
-    pub struct BuyCoreCollatorProof<T: Config> {
-        account: T::AccountId,
-        // TODO
-        _signature: (),
+    pub struct BuyCoreCollatorProof<PublicKey>
+    where
+        PublicKey: RuntimeAppPublic + Clone + core::fmt::Debug,
+    {
+        nonce: u64,
+        public_key: PublicKey,
+        signature: PublicKey::Signature,
+    }
+
+    impl<PublicKey> BuyCoreCollatorProof<PublicKey>
+    where
+        PublicKey: RuntimeAppPublic + Clone + core::fmt::Debug,
+    {
+        fn verify_signature(&self, para_id: ParaId) -> bool {
+            let payload = (self.nonce, para_id).encode();
+            self.public_key.verify(&payload, &self.signature)
+        }
+
+        #[cfg(any(test, feature = "runtime-benchmarks"))]
+        pub fn new(nonce: u64, para_id: ParaId, public_key: PublicKey) -> Option<Self> {
+            let payload = (nonce, para_id).encode();
+            public_key
+                .sign(&payload)
+                .map(|signature| BuyCoreCollatorProof {
+                    nonce,
+                    public_key,
+                    signature,
+                })
+        }
+
+        #[cfg(test)]
+        pub fn set_nonce(&mut self, nonce: u64) {
+            self.nonce = nonce;
+        }
+
+        #[cfg(test)]
+        pub fn set_public_key(&mut self, public_key: PublicKey) {
+            self.public_key = public_key;
+        }
+
+        #[cfg(test)]
+        pub fn set_signature(&mut self, signature: PublicKey::Signature) {
+            self.signature = signature;
+        }
```

Comment:
I would make the fields public instead of having setters

Reply:
I only want to make the fields freely accessible for testing. Can I do that? 

Reason being is that if you mutate this struct, it will become invalid as either public key or signature changes.

Reply:
For types that can be deserialized from bytes, such as this one that implements `Decode`, you must assume they can be in an invalid state anyway, so I don't see much benefit from having private fields.

Reply:
Oh. I did not meant it for validation. Verify function will return `false` in that case anyway.

My idea was to indicate anybody who is using this structure in non-test environment to be able to just call `new` and get valid instance which they cannot modify as any change there can render the struct invalid and call to fail. So, basically reduce the error surface for anybody working with the struct.

Reply:
@tmpolaczyk Now, the fields are public. :)

Reply:
Cool, thank you